packages/babel-plugin-proposal-object-rest-spread/src/index.js

```diff
@@ -260,6 +265,11 @@ export default declare((api, opts) => {
       if (!innerNode) {
         return false;
       }
+
+      // If this is an AssignmentPattern, always return true else the RHS will be duplicated
+      if (innerNode.type === "AssignmentPattern" && innerNode.left) {
+        return true;
+      }
```

Comment:
Similar to `RestElement`, we need to recurse into `node.left`.

```
innerNode ??= node.left
```

Reply:
So I originally added that but the output did not end up as I expected. Specifically the RHS was duplicated. I believe the way that we handle assignment patterns also duplicates the RHS. 

See this diff that I branched off this branch for: https://github.com/dan-kez/babel/pull/1/files

Reply:
Oh I was meant to replace the whole "if" branch with `innerNode ??= node.left`, which is similar to how we handle the array / object patterns (note that in your posted branch it was changed to `innerNode = innerNode.left`). The `innerNode` is essentially the arguments of next recursive call, so the algorithm can be summarized as

```
function hasMoreThanOneBinding(node) {
  if (node is ArrayPattern ) {
    return true if node has more than one elements
    else return hasMoreThanOneBinding(node's first element)
  } else if (node is ObjectPattern) {
    return true if node has more than one properties
    else return hasMoreThanOneBinding(node's first property)
  } else if (node is AssignmentPattern) {
    return hasMoreThanOneBinding(node's left)
  } else if (node is RestElement) {
    return hasMoreThanOneBinding(node's argument)
  }
}

Reply:
Ah I understand now. Apologies for all the back and forth here - I really appreciate you taking the time on this PR!

I've refactored the function to be more in line with you example. I did add one additional case `ObjectProperty`. This was because an ObjectPattern could have the first element a RestElement (for example `const { x15: { ...y15 } } = z();`). In this case `node.properties[0].value` would be undefined which would fall through the final else clause.

To be more specific I added an `ObjectProperty` case where I test on the presence of `node?.value` and then recurse on the given value.

If this is wrong please let me know - I'm happy to change it.


I also used optional chaining just in case I missed a case were node could be falsey (this was how I found the above issue when stepping through the test case). I am also happy to add an early return if node is undefined to return true or other cases that could be present in this LHS assignment.